common/persistence/sqlMetadataPersistence.go

```diff
@@ -0,0 +1,383 @@
+	"fmt"
+	_ "github.com/go-sql-driver/mysql" // MySQL driver
+	"github.com/jmoiron/sqlx"
+
+	"bytes"
+	"encoding/gob"
+	workflow "github.com/uber/cadence/.gen/go/shared"
+)
+
+type (
+	sqlMetadataManager struct {
+		db *sqlx.DB
+	}
+)
+
+const (
+	templateCreateDomainSqlQuery = `INSERT INTO domains (
+		id,
+		retention, 
+		emit_metric,
+		config_version,
+		name, 
+		status, 
+		description, 
+		owner_email,
+		failover_version, 
+		is_global_domain,
+		active_cluster_name, 
+		clusters
+		)
+		VALUES(
+		:domain_info_id,
+		:domain_config_retention, 
+		:domain_config_emit_metric,
+		:config_version,
+		:domain_info_name, 
+		:domain_info_status, 
+		:domain_info_description, 
+		:domain_info_owner_email,
+		:failover_version, 
+		:is_global_domain,
+		:domain_replication_config_active_cluster_name, 
+		:domain_replication_config_clusters
+		)`
+
+	getDomainPart = `SELECT
+		id,
+		retention, 
+		emit_metric,
+		config_version,
+		name, 
+		status, 
+		description, 
+		owner_email,
+		failover_version, 
+		is_global_domain,
+		active_cluster_name, 
+		clusters,
+		db_version
+FROM domains
+`
+	templateGetDomainByIdSqlQuery = getDomainPart +
+		`WHERE id = :id`
+	templateGetDomainByNameSqlQuery = getDomainPart +
+		`WHERE name = :name`
+
+	updateDomainPart = `UPDATE domains
+SET
+		id = :domain_info_id,
+		retention = :domain_config_retention, 
+		emit_metric = :domain_config_emit_metric,
+		config_version = :config_version,
+		name = :domain_info_name, 
+		status = :domain_info_status, 
+		description = :domain_info_description, 
+		owner_email = :domain_info_owner_email,
+		failover_version = :failover_version, 
+		active_cluster_name = :domain_replication_config_active_cluster_name,  
+		clusters = :domain_replication_config_clusters,
+		db_version = :next_db_version
+WHERE
+name = :domain_info_name
+AND
+`
+	templateUpdateDomainWhereCurrentVersionIsNullSqlQuery = updateDomainPart +
+		`db_version IS NULL`
+	templateUpdateDomainWhereCurrentVersionIsIntSqlQuery = updateDomainPart +
+		`db_version = :current_db_version`
+
+	templateDeleteDomainByIdSqlQuery   = `DELETE FROM domains WHERE id = :id`
+	templateDeleteDomainByNameSqlQuery = `DELETE FROM domains WHERE name = :name`
+)
+
+func (m *sqlMetadataManager) Close() {
+	if m.db != nil {
+		m.db.Close()
+	}
+}
+
+func gobSerialize(x interface{}) ([]byte, error) {
+	b := bytes.Buffer{}
+	e := gob.NewEncoder(&b)
+	err := e.Encode(x)
+	if err != nil {
+		return nil, err
+	}
+	return b.Bytes(), nil
+}
+
+func gobDeserialize(a []byte, x interface{}) error {
+	b := bytes.NewBuffer(a)
+	d := gob.NewDecoder(b)
+	err := d.Decode(x)
+	return err
+}
+
+func (m *sqlMetadataManager) CreateDomain(request *CreateDomainRequest) (*CreateDomainResponse, error) {
+	tx, err := m.db.Beginx()
+	if err != nil {
+		return nil, err
+	}
+	defer tx.Rollback()
+
+	// Disallow creating more than one domain with the same name, even if the UUID is different.
+	_, err = m.GetDomain(&GetDomainRequest{Name: request.Info.Name})
```

Comment:
check error first, then handle the actual logic outside.
if err == nil {
  return domain_already_exist;
}
if err!= nil and err!=not_exist {
 return error
}
// Continue with create flow.